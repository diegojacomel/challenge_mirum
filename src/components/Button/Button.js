// Modules
import React, { Fragment } from "react";
import PropTypes from 'prop-types';

// Components
import Icon from '../Icon/Icon';
import { Link } from 'react-router-dom';

// Styles
import styled from "styled-components";

const OutlineButton = (theme, outline, color) => {
    if (outline) {
        return `
            border: 1px solid ${color};
            background-color: ${theme.colors.white};
            color: ${color};
            &:hover {
                border: 1px solid ${theme.colors.white};
                background-color: ${color};
                color: ${theme.colors.white};
            }
    `;
    } else {
        return `
            background-color: ${color};
            color: ${theme.colors.white};
    `;
    }
};

const SizeButton = (props) => {
    switch (props.size) {
        case 'xg':
            return `
                padding: ${props.theme.spacing.lg} ${props.theme.spacing.xg};
            `
        case 'lg':
            return `
                padding: ${props.theme.spacing.md} ${props.theme.spacing.xlg};
            `
        case 'md':
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};
            `
        case 'sm':
            return `
                padding: ${props.theme.spacing.xsm} ${props.theme.spacing.md};
            `
        case 'xs':
            return `
                padding: ${props.theme.spacing.xs} ${props.theme.spacing.sm};
            `
        default:
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};'
        `
    }
}

const ButtonStyle = styled('button')`
    text-align: center;
    cursor: pointer;
    outline: none;
    border: none;
    line-height: 1;
    transition: all 0.3s;
    margin-right: 15px;
    white-space: nowrap;
    width: ${({ block }) => block ? '100%' : 'initial'};
    border-radius: ${({theme, rounded}) => theme.rounded[rounded]};
    font-size: ${({theme, size}) => theme.fontSize[size]};
    ${({ theme, outline, color }) => OutlineButton(theme, outline, theme.colors[color])};
    ${props => SizeButton(props)};
    &:hover {
        ${({ theme, outline, color }) => OutlineButton(theme, outline, theme.colors.hover[color])};
    }
    &:last-child {
        margin-right: 0;
    }
    & > span {
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-direction: ${({ iconLeft }) => iconLeft ? 'row-reverse' : 'row'};
        & > svg {
            margin-left: ${({ iconLeft }) => !iconLeft ? '10px' : '0'};
            margin-right: ${({ iconLeft }) => iconLeft ? '10px' : '0'};
        }
    }
`;

const LinkStyle = styled(Link)`
    display: inline-block;
    text-decoration: none;
    color: inherit;
    text-align: center;
    cursor: pointer;
    outline: none;
    border: none;
    line-height: 1;
    transition: all 0.3s;
    margin-right: 15px;
    white-space: nowrap;
    padding: 20px 30px;
    border-radius: 5px;
    font-size: 16px;
    color: #666666;
    border: 1px solid #999999;
    background: #eeeeee;
    &:hover {
        background: #999999;
        color: #ffffff;
    }
    &:last-child {
        margin-right: 0;
    }
    & > span {
        display: flex;
        justify-content: space-between;
        align-items: center;
    }
`

const Button = ({ children, type, color, outline, size, rounded, block, icon, iconSize, iconColor, iconLeft, disabled, onClick, linkTo }) => (
    <Fragment>
        {!!linkTo
            ?
            <LinkStyle id="testLinkButton" to={linkTo}>
                <span>
                    {children}
                    {icon
                        ?
                        <Icon tag={icon} iconSize={iconSize} iconColor={iconColor} />
                        :
                        null
                    }
                </span>
            </LinkStyle>
            :
            <ButtonStyle
                type={type}
                color={color}
                outline={outline}
                size={size}
                rounded={rounded}
                block={block}
                iconLeft={iconLeft}
                onClick={onClick}
                disabled={disabled}
                id="testButton"
            >
                <span>
                    {children}
                    {icon
                        ?
                        <Icon tag={icon} iconSize={iconSize} iconColor={iconColor} />
                        :
                        null
                    }
                </span>
            </ButtonStyle>
        }
    </Fragment>
);

Button.propTypes = {
    type: PropTypes.string,
    color: PropTypes.string,
    outline: PropTypes.bool,
    size: PropTypes.string,
    rounded: PropTypes.string,
    block: PropTypes.bool,
    icon: PropTypes.string,
    iconSize: PropTypes.string,
    iconColor: PropTypes.string,
    disabled: PropTypes.bool,
    linkTo: PropTypes.string,
    onClick: PropTypes.func
}

Button.defaultProps = {
    type: 'button',
    color: 'default',
    outline: false,
    size: 'md',
    rounded: 'sm',
    block: false,
    disabled: false,
    linkTo: null,
    onClick: () => console.log('clicked button')
}

export default Button;