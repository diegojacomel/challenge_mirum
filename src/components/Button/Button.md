## Basic usage

```js
<Button>Click here</Button>
```

## Props

#### **type**

The type attribute specifies the type of button.

type: string

#### **color**

It define the button color.

type: string

- primary
- danger
- info
- success
- default

```js
<div>
    <Button color="primary">primary</Button>
    <Button color="danger">danger</Button>
    <Button color="info">info</Button>
    <Button color="success">success</Button>
    <Button color="default">default</Button>
</div>
```

#### **outline**

It define if the button is outline or filled.

type: bool

```js
<div>
    <Button color="primary" outline>primary</Button>
    <Button color="danger" outline>danger</Button>
    <Button color="info" outline>info</Button>
    <Button color="success" outline>success</Button>
    <Button color="default" outline>default</Button>
</div>
```

#### **size**

It define the button size.

type: string

- xg
- lg
- md
- sm
- xs

```js
<div>
    <Button color="primary" size="xg">Button XG</Button>
    <Button color="primary" size="lg">Button LG</Button>
    <Button color="primary" size="md">Button MD</Button>
    <Button color="primary" size="sm">Button SM</Button>
    <Button color="primary" size="xs">Button XS</Button>
</div>
```

#### **rounded**

It define the button rounded size.

type: string

- none
- sm
- md
- lg
- full

```js
<Button color="primary" rounded="lg">Button Rounded</Button>
```

#### **block**

Let the button with full width display.

type: bool

```js
<Button color="primary" block>Button Block</Button>
```

#### **icon**

If you to wish putting a icon on button.

type: string

P.S.: To more information, you should find the Icon component on styleguide sidebar.

```js
<Button color="primary" icon="arrow-right2" iconSize="md" iconColor="white">
    Button Icon
</Button>
```

#### **iconSize**

It define the icon size.

type: string

- xxs
- xs
- xsm
- sm
- md
- lg
- xlg
- xg
- xxg

#### **iconColor**

It define the button color.

type: string

- primary
- danger
- info
- success
- default

#### **iconLeft**

It should be used when you want reverse the icon side.

type: bool

#### **onClick**

It should be used to fire any method that you want.

type: function