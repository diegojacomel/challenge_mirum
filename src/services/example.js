// Api
//import Api from '../utils/api';

class ExampleService {
    static products() {
        return new Promise((resolve) => {
            return resolve([{
                id: 0,
                checked: false,
                image: '',
                title: 'Insights',
                description: 'The best ecommerce overview in Latin America',
                value: 89.90
            },{
                id: 1,
                checked: false,
                image: '',
                title: 'Aliexpress',
                description: 'Pratical guide for to do excellent purchases in Aliexpress',
                value: 89.90
            },{
                id: 2,
                checked: false,
                image: '',
                title: 'Clothes sizes',
                description: 'Complete guide about clothes sizes from China',
                value: 89.90
            }])
        }, 3000)
    }
}

export default ExampleService;

// class RealService {
//     static getData(data) {
//         return Api.get(`parametro/${data}`);
//     }
// }

// export default RealService;