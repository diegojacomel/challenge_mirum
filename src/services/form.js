// Api
import Api from '../utils/api';

class FormService {
    static sendData(data) {
        return Api.post(`http://localhost:3000/register`, data);
    }
}

export default FormService;