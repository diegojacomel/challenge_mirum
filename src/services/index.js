// Services
import FormService from './form';

export {
    FormService
}