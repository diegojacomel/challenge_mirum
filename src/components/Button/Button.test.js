// Modules
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

// Theme
import { primaryTheme } from '../../styles/themes'; 

// Component
import Button from './Button';

// Services
import { ExampleService } from '../../services';

describe('Button', () => {
    const wrapper = shallow(<Button>Click here</Button>);

    test('It exist', () => {
        expect(wrapper.find('#testButton')).toExist();
    })

    test('At least one', () => {
        expect(wrapper).toContainMatchingElement('#testButton');
    })

    test('Have display name', () => {
        expect(wrapper.find('#testButton')).toHaveDisplayName('styled.button');
    })

    test('Match selector', () => {
        expect(wrapper.find('#testButton')).toMatchSelector('#testButton');
    })

    test('Have props', () => {
        expect(wrapper.find('#testButton')).toHaveProp('type');
        expect(wrapper.find('#testButton')).toHaveProp('color');
        expect(wrapper.find('#testButton')).toHaveProp('outline');
        expect(wrapper.find('#testButton')).toHaveProp('size');
        expect(wrapper.find('#testButton')).toHaveProp('rounded');
        expect(wrapper.find('#testButton')).toHaveProp('block');
        expect(wrapper.find('#testButton')).toHaveProp('disabled');
        expect(wrapper.find('#testButton')).toHaveProp('onClick');
    })

    it('Render correctly', () => {
        const tree = renderer.create(
            <ThemeProvider theme={primaryTheme}>
                <Button>
                    Click here
                </Button>
            </ThemeProvider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    })
    
    it('Test API', async () => {
        const data = await ExampleService.products();

        expect(data).toBeDefined()
    })
})