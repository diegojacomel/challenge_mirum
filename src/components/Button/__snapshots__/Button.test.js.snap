// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render correctly 1`] = `
<button
  className="sc-bwzfXH egfyjy"
  color="default"
  disabled={false}
  id="testButton"
  onClick={[Function]}
  size="md"
  type="button"
>
  <span>
    Click here
  </span>
</button>
`;
